livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -354,57 +375,107 @@ def _process_stream_event(self, data: dict, closing_ws: bool) -> None:
             logger.error("Received error from AssemblyAI: %s", data["error"])
             return
 
-        message_type = data.get("message_type")
+        message_type = data.get("type")
 
-        if message_type == "SessionBegins":
-            start_event = stt.SpeechEvent(type=stt.SpeechEventType.START_OF_SPEECH)
-            self._event_ch.send_nowait(start_event)
+        if message_type == "Begin":
+            logger.debug("AssemblyAI session started: %s", str(data))
 
-        elif message_type == "PartialTranscript":
-            alts = live_transcription_to_speech_data(ENGLISH, data)
-            if len(alts) > 0 and alts[0].text:
-                interim_event = stt.SpeechEvent(
-                    type=stt.SpeechEventType.INTERIM_TRANSCRIPT,
-                    alternatives=alts,
-                )
-                self._event_ch.send_nowait(interim_event)
-
-        elif message_type == "FinalTranscript":
+        elif message_type == "Turn":
+            logger.debug("AssemblyAI turn received: %s", str(data))
             alts = live_transcription_to_speech_data(ENGLISH, data)
-            if len(alts) > 0 and alts[0].text:
-                final_event = stt.SpeechEvent(
-                    type=stt.SpeechEventType.FINAL_TRANSCRIPT,
-                    alternatives=alts,
-                )
-                self._final_events.append(final_event)
-                self._event_ch.send_nowait(final_event)
-
-            # log metrics
-            if self._speech_duration > 0:
-                usage_event = stt.SpeechEvent(
-                    type=stt.SpeechEventType.RECOGNITION_USAGE,
-                    alternatives=[],
-                    recognition_usage=stt.RecognitionUsage(audio_duration=self._speech_duration),
-                )
-                self._event_ch.send_nowait(usage_event)
-                self._speech_duration = 0
-
-        elif message_type == "SessionTerminated":
+
+            if self._opts.token_mode:
+                # Get all words from the message
+                all_words = data.get("words", [])
+                end_of_turn = data.get("end_of_turn")
+            
+                if end_of_turn == True: 
+                    self._event_ch.send_nowait(stt.SpeechEvent(type=stt.SpeechEventType.END_OF_SPEECH)) 
+                
+                for i, word in enumerate(all_words):
+                    word_id = f"{data.get('turn_order', 0)}_{i}"  # Create a unique ID for this word position
+                    word_text = word.get("text", "")
+                
+                # Check if this is a new or updated word
+                if word_id in self._last_seen_words:
```

Comment:
This logic seems quite complex, is this intended by the AssemblyAI API to require that from users?